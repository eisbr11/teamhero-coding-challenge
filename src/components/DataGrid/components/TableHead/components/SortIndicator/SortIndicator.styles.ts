import { createUseStyles } from 'react-jss';

const useStyles = createUseStyles({
  indicator: {
    display: 'flex',
    flex: '0 0 16px',
  },
});

export default useStyles;
