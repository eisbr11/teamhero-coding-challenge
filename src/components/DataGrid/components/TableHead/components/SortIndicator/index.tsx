import { ActiveSort, ColConfig, SortDirection } from 'types/datagrid.type';
import { CaretDown, CaretUp } from 'components/CustomIcons';
import useStyles from './SortIndicator.styles';

const renderSvg = (direction: SortDirection) => {
  switch (direction) {
    case 'DESC':
      return <CaretUp />;
    case 'ASC':
      return <CaretDown />;
    case 'NONE':
    default:
      return null;
  }
};

const SortIndicator = ({
  column,
  activeSort,
}: {
  column: ColConfig,
  activeSort: ActiveSort,
}) => {
  const classes = useStyles();

  if (column.dataKey === activeSort.dataKey) {
    return (
      <span className={classes.indicator}>
        {renderSvg(activeSort.direction)}
      </span>
    );
  }
  return null;
};

export default SortIndicator;
