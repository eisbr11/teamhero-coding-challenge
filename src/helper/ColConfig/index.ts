import {
  Alignment,
  ColConfig,
  CompareFn, FilterFn, FilterType,
  Row,
} from 'types/datagrid.type';
import { compareStrings, filterByString } from 'helper/dataGrid.helper';

const createColConfig = ({
  dataKey,
  label,
  valueGetter,
  sortable = true,
  flexGrow = 0,
  sortFn = compareStrings,
  filterable = false,
  filterFn = filterByString,
  filterType = 'text',
  hide = false,
  align = 'left',
  width = 150,
}: {
  dataKey: string;
  label: string;
  valueGetter?: (row: Row, key: string) => React.ReactNode;
  sortable?: boolean,
  flexGrow?: number,
  sortFn?: CompareFn,
  filterable?: boolean,
  filterFn?: FilterFn,
  filterType?: FilterType,
  hide?: boolean,
  align?: Alignment,
  width?: number,
}): ColConfig => {
  const defaultValueGetter = (row: Row, key: string) => row[key];
  const configValueGetter = valueGetter || defaultValueGetter;
  return {
    dataKey,
    label,
    valueGetter: configValueGetter,
    sortable,
    flexGrow,
    sortFn,
    filterable,
    filterFn,
    filterType,
    hide,
    align,
    width,
  };
};

export default createColConfig;
