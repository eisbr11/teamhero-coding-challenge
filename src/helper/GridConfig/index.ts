import { GridConfig } from 'types/datagrid.type';

const createGridConfig = ({
  hideHead = false,
  showFilters = false,
}: {
  hideHead?: boolean,
  showFilters?: boolean,
}): GridConfig => ({
  hideHead,
  showFilters,
});

export default createGridConfig;
