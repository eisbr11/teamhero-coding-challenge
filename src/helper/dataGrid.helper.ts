import {
  ActiveFilter, ActiveSort,
  CellContent,
  CompareFn,
  FilterFn, Row,
  SortDirection,
} from 'types/datagrid.type';
import { Skill } from '../types/contacts.type';

/**
 * make Active Filters unique by key
 *
 * @param data
 * @param key
 */
const uniqueFilters = (data: ActiveFilter[], key: string = 'dataKey'): ActiveFilter[] => {
  const map = new Map();

  data.forEach((item: any) => {
    map.set(item[key], item);
  });

  return Array.from(map.values());
};

/**
 * filter function for filtering by string compare
 *
 * @param cellValue
 * @param filterValue
 */
const filterByString: FilterFn = (
  cellValue: CellContent,
  filterValue: string,
): boolean => {
  if (filterValue.length === 0) {
    return true;
  }
  return typeof cellValue === 'string' ? (
    cellValue?.toLocaleLowerCase().includes(filterValue.toLocaleLowerCase()))
    : true;
};

/**
 * filter function for filtering by tags
 *
 * @param cellValue
 * @param filterValue
 */
const filterByTags: FilterFn = (
  cellValue: CellContent,
  filterValue: string,
): boolean => {
  if (filterValue.length === 0) {
    return true;
  }
  const skills = cellValue as Skill[];

  const flatSkills = skills.map((skill) => skill.type);

  return flatSkills.includes(filterValue);
};

/**
 * compare Function for two strings
 */
const compareStrings: CompareFn = ((val1, val2) => {
  if (val1 && val2) {
    if (val1 < val2) {
      return -1;
    }
    if (val1 > val2) {
      return 1;
    }
  }
  return 0;
});

/**
 * update the sort direction Key depending on the old key
 */
const updateSortDirectionFromOldValue = ((oldDirection: SortDirection): SortDirection => {
  switch (oldDirection) {
    case 'NONE':
      return 'ASC';
    case 'ASC':
      return 'DESC';
    case 'DESC':
      return 'NONE';
    default:
      return 'ASC';
  }
});

const sortRows = (rows: Row[], activeSort: ActiveSort): Row[] => {
  switch (activeSort.direction) {
    case 'ASC':
      return rows.sort((a: Row, b: Row) => activeSort.sortFn(
        a[activeSort.dataKey], b[activeSort.dataKey],
      ));
    case 'DESC':
      return rows.sort((a: Row, b: Row) => activeSort.sortFn(
        a[activeSort.dataKey], b[activeSort.dataKey],
      )).reverse();
    default:
      return rows;
  }
};

export {
  uniqueFilters,
  filterByString,
  filterByTags,
  compareStrings,
  updateSortDirectionFromOldValue,
  sortRows,
};
