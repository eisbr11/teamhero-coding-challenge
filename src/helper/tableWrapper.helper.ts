import { ColConfig } from 'types/datagrid.type';

const calculateMinWidth = (columns: ColConfig[]): number => {
  let minWidth = 0;

  columns.forEach((column) => {
    if (!column.hide) {
      minWidth += column.width;
    }
  });

  return minWidth;
};

export default calculateMinWidth;
