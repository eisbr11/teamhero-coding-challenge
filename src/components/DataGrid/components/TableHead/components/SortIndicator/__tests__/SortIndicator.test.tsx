import React from 'react';
import { ThemeProvider } from 'react-jss';

import { mount } from 'enzyme';
import teamHeroTheme from 'themes/teamHero.theme';
import { ActiveSort, ColConfig } from 'types/datagrid.type';
import createColConfig from 'helper/ColConfig';
import { compareStrings } from 'helper/dataGrid.helper';
import SortIndicator from '../index';

describe('Sort Indicator', () => {
  const column: ColConfig = createColConfig({
    dataKey: 'someOtherKey',
    label: 'LABEL',
    sortable: true,
  });

  it('should render correctly with sort Ascending', () => {
    const activeSortAsc:ActiveSort = {
      direction: 'ASC',
      sortFn: compareStrings,
      dataKey: 'someOtherKey',
    };

    const sortIndicator = mount(
      <ThemeProvider theme={teamHeroTheme}>
        <SortIndicator column={column} activeSort={activeSortAsc} />
      </ThemeProvider>,
    );

    expect(sortIndicator).toMatchSnapshot();
  });

  it('should render correctly with sort Ascending', () => {
    const activeSortDesc:ActiveSort = {
      direction: 'DESC',
      sortFn: compareStrings,
      dataKey: 'someOtherKey',
    };

    const sortIndicator = mount(
      <ThemeProvider theme={teamHeroTheme}>
        <SortIndicator column={column} activeSort={activeSortDesc} />
      </ThemeProvider>,
    );

    expect(sortIndicator).toMatchSnapshot();
  });

  it('should render correctly with None sort', () => {
    const activeSortNone:ActiveSort = {
      direction: 'NONE',
      sortFn: compareStrings,
      dataKey: 'someOtherKey',
    };

    const sortIndicator = mount(
      <ThemeProvider theme={teamHeroTheme}>
        <SortIndicator column={column} activeSort={activeSortNone} />
      </ThemeProvider>,
    );

    expect(sortIndicator).toMatchSnapshot();
  });

  it('should render correctly with no active sort associated', () => {
    const activeSortNonAssociated:ActiveSort = {
      direction: 'ASC',
      sortFn: compareStrings,
      dataKey: 'different',
    };

    const sortIndicator = mount(
      <ThemeProvider theme={teamHeroTheme}>
        <SortIndicator column={column} activeSort={activeSortNonAssociated} />
      </ThemeProvider>,
    );

    expect(sortIndicator).toMatchSnapshot();
  });
});
