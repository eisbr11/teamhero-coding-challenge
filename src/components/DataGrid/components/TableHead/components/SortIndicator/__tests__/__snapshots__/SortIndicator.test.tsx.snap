// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort Indicator should render correctly with None sort 1`] = `
<ThemeProvider
  theme={
    Object {
      "colorBlack": "#000000",
      "colorGrayDark": "#666666",
      "colorGrayLight": "#f1f4f9",
      "colorHighlight": "#fef0d8",
      "colorPrimary": "#d51641",
      "colorSecondary": "#a3b415",
      "colorWhite": "#FFFFFF",
    }
  }
>
  <SortIndicator
    activeSort={
      Object {
        "dataKey": "someOtherKey",
        "direction": "NONE",
        "sortFn": [Function],
      }
    }
    column={
      Object {
        "align": "left",
        "dataKey": "someOtherKey",
        "filterFn": [Function],
        "filterTags": Array [],
        "filterType": "text",
        "filterable": false,
        "flexGrow": 0,
        "hide": false,
        "label": "LABEL",
        "sortFn": [Function],
        "sortable": true,
        "valueGetter": [Function],
        "width": 150,
      }
    }
  >
    <span
      className="indicator-0-2-1"
    />
  </SortIndicator>
</ThemeProvider>
`;

exports[`Sort Indicator should render correctly with no active sort associated 1`] = `
<ThemeProvider
  theme={
    Object {
      "colorBlack": "#000000",
      "colorGrayDark": "#666666",
      "colorGrayLight": "#f1f4f9",
      "colorHighlight": "#fef0d8",
      "colorPrimary": "#d51641",
      "colorSecondary": "#a3b415",
      "colorWhite": "#FFFFFF",
    }
  }
>
  <SortIndicator
    activeSort={
      Object {
        "dataKey": "different",
        "direction": "ASC",
        "sortFn": [Function],
      }
    }
    column={
      Object {
        "align": "left",
        "dataKey": "someOtherKey",
        "filterFn": [Function],
        "filterTags": Array [],
        "filterType": "text",
        "filterable": false,
        "flexGrow": 0,
        "hide": false,
        "label": "LABEL",
        "sortFn": [Function],
        "sortable": true,
        "valueGetter": [Function],
        "width": 150,
      }
    }
  />
</ThemeProvider>
`;

exports[`Sort Indicator should render correctly with sort Ascending 1`] = `
<ThemeProvider
  theme={
    Object {
      "colorBlack": "#000000",
      "colorGrayDark": "#666666",
      "colorGrayLight": "#f1f4f9",
      "colorHighlight": "#fef0d8",
      "colorPrimary": "#d51641",
      "colorSecondary": "#a3b415",
      "colorWhite": "#FFFFFF",
    }
  }
>
  <SortIndicator
    activeSort={
      Object {
        "dataKey": "someOtherKey",
        "direction": "ASC",
        "sortFn": [Function],
      }
    }
    column={
      Object {
        "align": "left",
        "dataKey": "someOtherKey",
        "filterFn": [Function],
        "filterTags": Array [],
        "filterType": "text",
        "filterable": false,
        "flexGrow": 0,
        "hide": false,
        "label": "LABEL",
        "sortFn": [Function],
        "sortable": true,
        "valueGetter": [Function],
        "width": 150,
      }
    }
  >
    <span
      className="indicator-0-2-1"
    >
      <CaretDown
        className=""
      >
        <svg
          aria-hidden="true"
          className=""
          viewBox="0 0 24 24"
        >
          <path
            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
          />
        </svg>
      </CaretDown>
    </span>
  </SortIndicator>
</ThemeProvider>
`;

exports[`Sort Indicator should render correctly with sort Ascending 2`] = `
<ThemeProvider
  theme={
    Object {
      "colorBlack": "#000000",
      "colorGrayDark": "#666666",
      "colorGrayLight": "#f1f4f9",
      "colorHighlight": "#fef0d8",
      "colorPrimary": "#d51641",
      "colorSecondary": "#a3b415",
      "colorWhite": "#FFFFFF",
    }
  }
>
  <SortIndicator
    activeSort={
      Object {
        "dataKey": "someOtherKey",
        "direction": "DESC",
        "sortFn": [Function],
      }
    }
    column={
      Object {
        "align": "left",
        "dataKey": "someOtherKey",
        "filterFn": [Function],
        "filterTags": Array [],
        "filterType": "text",
        "filterable": false,
        "flexGrow": 0,
        "hide": false,
        "label": "LABEL",
        "sortFn": [Function],
        "sortable": true,
        "valueGetter": [Function],
        "width": 150,
      }
    }
  >
    <span
      className="indicator-0-2-1"
    >
      <CaretUp
        className=""
      >
        <svg
          aria-hidden="true"
          className=""
          viewBox="0 0 24 24"
        >
          <path
            d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
          />
        </svg>
      </CaretUp>
    </span>
  </SortIndicator>
</ThemeProvider>
`;
